astronomy/standard_product_3.cpp

```diff
@@ -309,30 +366,76 @@ StandardProduct3::StandardProduct3(
         }
       }
 
-      orbit.Append(epoch, {position, velocity});
+      // Bad or absent positional and velocity values are to be set to 0.000000.
+      if (position == ITRS::origin || velocity == Velocity<ITRS>()) {
+        if (!arc.Empty()) {
+          orbit.push_back(make_not_null_unique<DiscreteTrajectory<ITRS>>());
+        }
+      } else {
+        arc.Append(epoch, {position, velocity});
+      }
     }
   }
   if (dialect != Dialect::ILRSA) {
     CHECK_EQ(columns(1, 3), "EOF") << location;
     read_line();
   }
+  for (auto& [id, orbit] : orbits_) {
+    // Do not leave a final empty trajectory if the orbit ends with missing
+    // data.
+    if (orbit.back()->Empty()) {
+      orbit.pop_back();
+    }
+  }
   CHECK(!line.has_value()) << location;
+  if (!has_velocities_) {
+    for (auto& [id, orbit] : orbits_) {
+      for (auto& arc : orbit) {
+        switch (arc->Size()) {
+#define COMPUTE_VELOCITIES_CASE(n)    \
```

Comment:
Align the `\` (clang-format should do it).

Also, preference for *not* (un)defining macros in the middle of statements.

Reply:
Done.